### PR TITLE
pause and start experiment by API

### PR DESCRIPTION
**Pull request by cwen0**
Signed-off-by: cwen0 <cwenyin0@gmail.com>

### What problem does this PR solve? 

#370 

We need to support pause and start chaos experiment by API.

### CODE REVIEW

File: api/v1alpha1/kinds.go [hunk 1]
```diff
@@ -13,34 +13,50 @@
...
+
+func (c *chaosKindMap) clone() map[string]*ChaosKind {
+	c.RLock()
+	defer c.RUnlock()
+
+	out := make(map[string]*ChaosKind)
+	for key, kind := range c.kinds {
+		out[key] = &ChaosKind{
+			Chaos:     kind.Chaos.DeepCopyObject(),
+			ChaosList: kind.ChaosList,
```
Comment (yeya24) [inline api/v1alpha1/kinds.go:44]: I am not sure whether `DeepCopyObject()` is necessary or not. If it is, ChaosList also supports this method. `kind.ChaosList.DeepCopyObject()`

File: pkg/apiserver/experiment/experiment.go [hunk 2]
```diff
@@ -600,3 +608,109 @@ func (s *Service) state(c *gin.Context) {
...
+// @Summary Start the paused chaos experiment by API
+// @Description Start the paused chaos experiment by API
+// @Tags experiments
+// @Produce json
+// @Param kind path string true "kind"
+// @Param namespace path string true "namespace"
+// @Param name path string true "name"
+// @Success 200 "start ok"
+// @Failure 500 {object} utils.APIError
+// @Failure 404 {object} utils.APIError
```
Comment (yeya24) [inline pkg/apiserver/experiment/experiment.go:664]: 400 Failure

File: pkg/apiserver/experiment/experiment.go [hunk 3]
```diff
@@ -600,3 +608,109 @@ func (s *Service) state(c *gin.Context) {
...
+// @Router /api/experiments/start/{kind}/{ns}/{name} [put]
```
Comment (yeya24) [inline pkg/apiserver/experiment/experiment.go:665]:
```
// @Router /api/experiments/start/{kind}/{namespace}/{name} [put]
```

File: pkg/apiserver/experiment/experiment.go [hunk 4]
```diff
@@ -600,3 +608,109 @@ func (s *Service) state(c *gin.Context) {
...
+// @Summary Pause chaos experiment by API
+// @Description Pause chaos experiment by API
+// @Tags experiments
+// @Produce json
+// @Param kind path string true "kind"
+// @Param namespace path string true "namespace"
+// @Param name path string true "name"
+// @Success 200 "pause ok"
+// @Failure 500 {object} utils.APIError
+// @Failure 404 {object} utils.APIError
```
Comment (yeya24) [inline pkg/apiserver/experiment/experiment.go:628]: 400 Failure is also needed

File: pkg/apiserver/experiment/experiment.go [hunk 5]
```diff
@@ -600,3 +608,109 @@ func (s *Service) state(c *gin.Context) {
 
 	c.JSON(http.StatusOK, data)
 }
+
+// ExperimentBase is used to identify the unique experiment from API request.
+type ExperimentBase struct {
+	Kind      string `uri:"kind" binding:"required,oneof=PodChaos NetworkChaos IoChaos StressChaos TimeChaos"`
```
Comment (yeya24) [inline pkg/apiserver/experiment/experiment.go:614]: `KernelChaos` :D

File: pkg/apiserver/experiment/experiment.go [hunk 6]
```diff
@@ -81,8 +84,10 @@ func Register(r *gin.RouterGroup, s *Service) {
 	// TODO: add more api handlers
 	endpoint.GET("", s.listExperiments)
 	endpoint.POST("/new", s.createExperiment)
-	endpoint.GET("/detail/:namespace/:name", s.getExperimentDetail)
-	endpoint.DELETE("/:kind/:namespace/:name", s.deleteExperiment)
+	endpoint.GET("/detail/:kind/:namespace/:name", s.getExperimentDetail)
+	endpoint.DELETE("/delete/:kind/:namespace/:name", s.deleteExperiment)
```
Comment (yeya24) [inline pkg/apiserver/experiment/experiment.go:88]: IMO `/delete/:kind/:namespace/:name` is not needed, do you mind changing this to `/:kind/:namespace/:name`? We are already using the `DELETE` verb here.

File: api/v1alpha1/kinds.go [hunk 7]
```diff
@@ -13,34 +13,50 @@
...
-		KindNetworkChaos: {Chaos: &NetworkChaos{}, ChaosList: &NetworkChaosList{}},
-		KindIOChaos:      {Chaos: &IoChaos{}, ChaosList: &IoChaosList{}},
-		KindKernelChaos:  {Chaos: &KernelChaos{}, ChaosList: &KernelChaosList{}},
-		KindTimeChaos:    {Chaos: &TimeChaos{}, ChaosList: &TimeChaosList{}},
-		KindStressChaos:  {Chaos: &StressChaos{}, ChaosList: &StressChaosList{}},
+// +kubebuilder:object:generate=false
+
+// ChaosKindMap defines a map including all chaos kinds.
+type chaosKindMap struct {
+	sync.RWMutex
```
[the remark this answers is not in the document]
Comment (cwen0) [reply api/v1alpha1/kinds.go:26]: This is a map, so I think it is necessary

File: api/v1alpha1/kinds.go (hunk 1 above)
Comment (cwen0) [reply api/v1alpha1/kinds.go:44]: OK, `DeepCopyObject()` is not necessary and I have removed it.

**Pull request merged**